clients/src/test/java/org/apache/kafka/clients/producer/internals/BufferPoolTest.java

```diff
@@ -377,4 +378,58 @@ public void run() {
         }
     }
 
+    @Test
+    public void testCloseAllocations() throws Exception {
+        BufferPool pool = new BufferPool(10, 1, metrics, Time.SYSTEM, metricGroup);
+        ByteBuffer buffer = pool.allocate(1, maxBlockTimeMs);
+
+        // Close the buffer pool. This should prevent any further allocations.
+        pool.close();
+
+        try {
+            pool.allocate(1, maxBlockTimeMs);
+            fail("Should have thrown KafkaException");
+        } catch (KafkaException e) {
+            // Expected.
+        }
+
+        // Ensure deallocation still works.
+        pool.deallocate(buffer);
+    }
+
+    @Test
+    public void testCloseNotifyWaiters() throws Exception {
+        BufferPool pool = new BufferPool(10, 1, metrics, Time.SYSTEM, metricGroup);
+        ByteBuffer buffer = pool.allocate(10, Long.MAX_VALUE);
+
+        CountDownLatch waiter1 = asyncAllocateClose(pool, 10);
+        CountDownLatch waiter2 = asyncAllocateClose(pool, 10);
+
+        assertEquals("Allocation shouldn't have happened yet, waiting on memory", 2L, waiter1.getCount() + waiter2.getCount());
+
+        // Close the buffer pool. This should notify all waiters.
+        pool.close();
+
+        assertTrue("Allocation should fail soon after close", waiter1.await(1, TimeUnit.SECONDS) && waiter2.await(1, TimeUnit.SECONDS));
+
+        pool.deallocate(buffer);
+    }
+
+    private CountDownLatch asyncAllocateClose(final BufferPool pool, final int size) {
+        final CountDownLatch completed = new CountDownLatch(1);
+        Thread thread = new Thread() {
+            public void run() {
+                try {
+                    pool.allocate(size, maxBlockTimeMs);
+                    fail("Unexpected allocation");
+                } catch (KafkaException e) {
+                    completed.countDown();
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
```

Comment:
If we used an executor with callables, we could propagate this directly to a future, which could be awaited in the test case.

Reply:
Done. Please verify I have it simplified.